### PR TITLE
Fix mypy errors attributed to `pytorch_lightning.trainer.connectors.callback_connector.py`

### LINKED ISSUE #13445: Improve typing coverage (4/n)
## 🚀 Typing coverage

Let's improve typing coverage of PyTorch Lightning together!

I'm creating a new issue in order to increase visibility. There are three older issues (#7037, #5023, #4698) which became stale over time.

### Plan

Currently, there are 55 files which are excluded from mypy checks so that our CI does not fail. These files vastly differ in difficulty in order to make the typing complete. For this reason, we are introducing difficulty estimate for each file so that community members can choose to work on the files appropriate to their skill level.

Please, comment on this issue in order to reserve a particular file to work on. Once you do so, I will edit this top comment to avoid collisions. Once you think your work is finished, please open a PR referencing this issue which:

 * removes the corresponding line from `pyproject.toml`
 * and passes `mypy` checks with the corresponding line removed. You can test it locally by running `mypy` from root directory

If you are struggling with pushing it over the finish line, open the PR anyway and someone from our team will help you to get it there. 🚀

Please note, that it can happen that you may need to edit more than just one file. This is fine, but please keep in mind, that the goal of your PR will be to make the check passing for the chosen file. Also, please note that the difficulty is just an educated guess.

For those of you who are not familiar with the process of contributing a PR, we have prepared a [simple guide](https://devblog.pytorchlightning.ai/quick-contribution-guide-86d977171b3a) that will walk you through the necessary steps. You can do it! :rocket: :muscle: 

#### Difficulty 3 of 3

- [x] `pytorch_lightning/callbacks/progress/rich_progress.py` @donlapark #14963 

### Completed

<details>

#### Difficulty 1 of 3

- [x] `pytorch_lightning/core/decorators.py` #14044
- [x] `pytorch_lightning/profilers/advanced.py` @nninept #13792 
~- [ ] `pytorch_lightning/profilers/base.py` @LeeChanHyuk #13879~
- [x] `pytorch_lightning/loggers/base.py` @JustinGoheen #13494 
- [x] `pytorch_lightning/__setup__.py` @CyprienRicque #13472 
~- [ ] `pytorch_lightning/distributed/dist.py` @puhuk #13492~
- [x] `pytorch_lightning/strategies/single_device.py` @CyprienRicque #13532 
- [x] `pytorch_lightning/trainer/optimizers.py` @gautierdag #13470
- [x] `pytorch_lightning/utilities/distributed.py` @krishnakalyan3 #13678
- [x] `pytorch_lightning/callbacks/finetuning.py` @ar90n #13516 
- [x] `pytorch_lightning/loggers/mlflow.py` @JustinGoheen ~~#13690~~ #13691
- [x] `pytorch_lightning/tuner/tuning.py` @donlapark ~~#13616~~ #13631 
- [x] `pytorch_lightning/strategies/single_tpu.py` @CyprienRicque #13534 
- [x] `pytorch_lightning/strategies/ddp2.py` @CyprienRicque #13535 
- [x] `pytorch_lightning/strategies/parallel.py` @CyprienRicque #13556 
- [x] `pytorch_lightning/loggers/csv_logs.py` @JustinGoheen #13538 
- [x] `pytorch_lightning/tuner/lr_finder.py` @donlapark #13513 #13652 
- [x] `pytorch_lightning/strategies/dp.py` @CyprienRicque #13564 
- [x] `pytorch_lightning/profilers/simple.py` @krishnakalyan3 #14103
- [x] `pytorch_lightning/strategies/sharded_spawn.py` @krishnakalyan3 #14102
- [x] `pytorch_lightning/demos/mnist_datamodule.py` @alro923 #13929 
- [x] `pytorch_lightning/demos/boring_classes.py` @krishnakalyan3 #14201
- [x] `pytorch_lightning/tuner/batch_size_scaling.py` @ar90n #13518 

#### Difficulty 2 of 3

- [x] `pytorch_lightning/loops/epoch/training_epoch_loop.py` @himkt #13555 
- [x] `pytorch_lightning/core/mixins/device_dtype_mixin.py` @krishnakalyan3 #13704 
- [x] `pytorch_lightning/loggers/comet.py` @JustinGoheen #13689 
- [x] `pytorch_lightning/loggers/tensorboard.py` @JustinGoheen #13688 
- [x] `pytorch_lightning/strategies/horovod.py` @CyprienRicque #13570 
- [x] `pytorch_lightning/callbacks/model_checkpoint.py` @BongYang #13617
- [x] `pytorch_lightning/strategies/fully_sharded.py` @BongYang #13941 
- [x] `pytorch_lightning/loggers/neptune.py` @JustinGoheen #13692 
- [x] `pytorch_lightning/utilities/meta.py` @nninept #13763 #13868
- [x] `pytorch_lightning/strategies/tpu_spawn.py` @BongYang #13813  
- [x] `pytorch_lightning/loggers/logger.py` @JustinGoheen #13541 
- [x] `pytorch_lightning/loggers/wandb.py` @gautierdag #13483
- [x] `pytorch_lightning/callbacks/stochastic_weight_avg.py` @donlapark #13685 #13860
- [x] `pytorch_lightning/strategies/strategy.py` @CyprienRicque #13519 
- [x] `pytorch_lightning/strategies/deepspeed.py` @donlapark #13832 
- [x] `pytorch_lightning/strategies/ddp_spawn.py` @donlapark #13865 
- [x] `pytorch_lightning/strategies/ipu.py` @HalestormAI #13786
- [x] `pytorch_lightning/trainer/connectors/callback_connector.py` @krishnakalyan3 #13750
- [x] `pytorch_lightning/strategies/ddp.py` @lijm1358 #13885 
- [x] `pytorch_lightning/core/saving.py` @JustinGoheen #13932 
- [x] `pytorch_lightning/callbacks/quantization.py` @krishnakalyan3 #13782 
- [x] `pytorch_lightning/strategies/sharded.py` @lijm1358 #14184 
- [x] `pytorch_lightning/core/datamodule.py` @JustinGoheen #13693

#### Difficulty 3 of 3
~- [ ] `pytorch_lightning/trainer/callback_hook.py` @JustinGoheen #13807 ~
- [x] `pytorch_lightning/core/module.py` @JustinGoheen #13603 
- [x] `pytorch_lightning/trainer/connectors/data_connector.py` @JustinGoheen #13806
- [x] `pytorch_lightning/utilities/auto_restart.py` @donlapark #13904 
- [x] `pytorch_lightning/trainer/supporters.py` @donlapark #14633 
- [x] `pytorch_lightning/profilers/pytorch.py` @krishnakalyan3 #14405 
- [x] `pytorch_lightning/utilities/data.py` @nandwalritik #13901 
- [x] `pytorch_lightning/trainer/trainer.py` ~@JustinGoheen #13810~ @BongYang #14204 

</details>

cc @borda @justusschock @awaelchli @rohitgr7 @Borda @tchaton @aniketmaurya @kingjuno @alat-rights @carmocca @akihironitta @stancld as you were all involved in previous issues

### PR DESCRIPTION
**Pull request by krishnakalyan3**
Part of #13445

### CODE REVIEW

File: src/pytorch_lightning/trainer/connectors/callback_connector.py [hunk 1]
```diff
@@ -243,8 +246,8 @@ def _reorder_callbacks(callbacks: List[Callback]) -> List[Callback]:
             A new list in which the last elements are Checkpoint if there were any present in the
             input.
         """
-        checkpoints = [c for c in callbacks if isinstance(c, Checkpoint)]
-        not_checkpoints = [c for c in callbacks if not isinstance(c, Checkpoint)]
+        checkpoints = [c for c in callbacks if isinstance(c, Checkpoint and Callback)]
```
Comment (otaj) [inline src/pytorch_lightning/trainer/connectors/callback_connector.py:249]: First of all, this is not how you check for multiple types - and we don't want to check for callback.

```suggestion
        checkpoints: List[Callback] = [c for c in callbacks if isinstance(c, Checkpoint)]
```

File: src/pytorch_lightning/trainer/connectors/callback_connector.py [hunk 2]
```diff
@@ -243,8 +246,8 @@ def _reorder_callbacks(callbacks: List[Callback]) -> List[Callback]:
...
+        not_checkpoints = [c for c in callbacks if not isinstance(c, Checkpoint and Callback)]
```
Comment (otaj) [inline src/pytorch_lightning/trainer/connectors/callback_connector.py:250]:
```suggestion
        not_checkpoints = [c for c in callbacks if not isinstance(c, Checkpoint)]
```

File: src/pytorch_lightning/trainer/connectors/callback_connector.py [hunk 3]
```diff
@@ -265,7 +268,7 @@ def _configure_external_callbacks() -> List[Callback]:
 
         factories = iter_entry_points("pytorch_lightning.callbacks_factory")
 
-    external_callbacks = []
+    external_callbacks: List[Callback] = []
     for factory in factories:
         callback_factory = factory.load()
         callbacks_list: List[Callback] = callback_factory()
```
Comment (otaj) [inline src/pytorch_lightning/trainer/connectors/callback_connector.py:274]:
```suggestion
        callbacks_list: Union[List[Callback], Callback] = callback_factory()
```

**Review by otaj — COMMENTED** [review]
There are still couple parts that are not working well for mypy, even with applied suggestions. In the method `_configure_accumulated_gradients`, this could be solved by differentiating between `grad_accum_callbacks` and `grad_accum_callback` (i.e. two separate variables, one for the case when it's a list and second for when it's a single instance)

**Comment by otaj** [comment]
I think that `src/pytorch_lightning/trainer/connectors/callback_connector.py:269: error: Incompatible types in assignment (expression has type "Generator[pkg_resources.EntryPoint, None, None]", variable has type "Union[List[importlib.metadata.EntryPoint], Tuple[]]")  [assignment]` will have to be silenced, because there's no reasonable fix to it

**Comment by krishnakalyan3** [comment]
Thank you for the reviews @otaj

File: src/pytorch_lightning/trainer/connectors/callback_connector.py [hunk 4]
```diff
@@ -148,10 +151,10 @@ def _configure_model_summary_callback(self, enable_model_summary: bool) -> None:
         progress_bar_callback = self.trainer.progress_bar_callback
         is_progress_bar_rich = isinstance(progress_bar_callback, RichProgressBar)
 
+        model_summary: ModelSummary = ModelSummary()
```
Comment (otaj) [inline src/pytorch_lightning/trainer/connectors/callback_connector.py:154]: Don't assign here, it's only for purpose of typing
```suggestion
        model_summary: ModelSummary
```

File: src/pytorch_lightning/trainer/connectors/callback_connector.py [hunk 5]
```diff
@@ -95,16 +96,18 @@ def on_trainer_init(
...
-        if grad_accum_callback:
+        if grad_accum_callbacks:
             if accumulate_grad_batches is not None:
                 raise MisconfigurationException(
                     "You have set both `accumulate_grad_batches` and passed an instance of "
                     "`GradientAccumulationScheduler` inside callbacks. Either remove `accumulate_grad_batches` "
                     "from trainer or remove `GradientAccumulationScheduler` from callbacks list."
                 )
-            grad_accum_callback = grad_accum_callback[0]
+            grad_accum_callback: Callback = grad_accum_callbacks[0]
```
Comment (otaj) [inline src/pytorch_lightning/trainer/connectors/callback_connector.py:110]: nit: Maybe do the same in here with declaring the variable before the if (but only declaring type, not assigning anything)

**Review by otaj — APPROVED** [review]

File: src/pytorch_lightning/trainer/connectors/callback_connector.py [hunk 6]
```diff
@@ -95,16 +96,19 @@ def on_trainer_init(
...
         self, accumulate_grad_batches: Optional[Union[int, Dict[int, int]]] = None
     ) -> None:
-        grad_accum_callback = [cb for cb in self.trainer.callbacks if isinstance(cb, GradientAccumulationScheduler)]
+        grad_accum_callbacks: List[Callback] = [
+            cb for cb in self.trainer.callbacks if isinstance(cb, GradientAccumulationScheduler)
+        ]
 
-        if grad_accum_callback:
+        if grad_accum_callbacks:
+            grad_accum_callback: Callback
```
Comment (otaj) [inline src/pytorch_lightning/trainer/connectors/callback_connector.py:104]: Very small nit: In previous comment, I meant to move the typing even outside if :D

```suggestion
        grad_accum_callback: Callback
        if grad_accum_callbacks:
```

**Review by otaj — APPROVED** [review]

**Review by otaj — APPROVED** [review]

File: src/pytorch_lightning/trainer/connectors/callback_connector.py [hunk 7]
```diff
@@ -243,7 +248,7 @@ def _reorder_callbacks(callbacks: List[Callback]) -> List[Callback]:
             A new list in which the last elements are Checkpoint if there were any present in the
             input.
         """
-        checkpoints = [c for c in callbacks if isinstance(c, Checkpoint)]
+        checkpoints: List[Callback] = [c for c in callbacks if isinstance(c, Checkpoint)]
```
Comment (rohitgr7) [inline src/pytorch_lightning/trainer/connectors/callback_connector.py:251]:
```suggestion
        checkpoints: List[Checkpoint] = [c for c in callbacks if isinstance(c, Checkpoint)]
```

File: src/pytorch_lightning/trainer/connectors/callback_connector.py [hunk 8]
```diff
@@ -95,16 +96,19 @@ def on_trainer_init(
     def _configure_accumulated_gradients(
         self, accumulate_grad_batches: Optional[Union[int, Dict[int, int]]] = None
     ) -> None:
-        grad_accum_callback = [cb for cb in self.trainer.callbacks if isinstance(cb, GradientAccumulationScheduler)]
+        grad_accum_callbacks: List[Callback] = [
```
Comment (rohitgr7) [inline src/pytorch_lightning/trainer/connectors/callback_connector.py:99]:
```suggestion
        grad_accum_callbacks: List[GradientAccumulationScheduler] = [
```

Comment (rohitgr7) [reply src/pytorch_lightning/trainer/connectors/callback_connector.py:99]: possibly a little more explicit.. Hope it works

**Review by rohitgr7 — APPROVED** [review]

File: src/pytorch_lightning/trainer/connectors/callback_connector.py [hunk 9]
```diff
@@ -95,16 +96,19 @@ def on_trainer_init(
...
+            cb for cb in self.trainer.callbacks if isinstance(cb, GradientAccumulationScheduler)
+        ]
 
-        if grad_accum_callback:
+        grad_accum_callback: Callback
```
Comment (awaelchli) [inline src/pytorch_lightning/trainer/connectors/callback_connector.py:103]:
```suggestion
```
Pretty sure it is not necessary.

**Review by awaelchli — APPROVED** [review]

File: src/pytorch_lightning/trainer/connectors/callback_connector.py (hunk 7 above)
Comment (awaelchli) [reply src/pytorch_lightning/trainer/connectors/callback_connector.py]: I saw an error when adding the two list together, mypy wants the types to match.

File: src/pytorch_lightning/trainer/connectors/callback_connector.py [hunk 10]
```diff
@@ -243,7 +247,7 @@ def _reorder_callbacks(callbacks: List[Callback]) -> List[Callback]:
             A new list in which the last elements are Checkpoint if there were any present in the
             input.
         """
-        checkpoints = [c for c in callbacks if isinstance(c, Checkpoint)]
+        checkpoints: List[Checkpoint] = [c for c in callbacks if isinstance(c, Checkpoint)]
```
Comment (rohitgr7) [inline src/pytorch_lightning/trainer/connectors/callback_connector.py:250]:
```suggestion
        checkpoints: List[Callback] = [c for c in callbacks if isinstance(c, Checkpoint)]
```

**Review by Borda — APPROVED** [review]

**Pull request merged**